File: routes/favorites.py
# ============================================================
# routes/favorites.py —— 收藏功能模块
# ============================================================

from flask import (
    Blueprint,
    render_template,
    request,
    redirect,
    session,
    jsonify
)
from db import get_connection
from utils import get_pagination   # 导入分页工具

favorites_bp = Blueprint('favorites', __name__)


@favorites_bp.route('/favorites')
def favorites():
    if 'user_id' not in session:
        return redirect('/login')

    per_page = 10
    user_id = session['user_id']

    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("SELECT COUNT(*) FROM Favorites WHERE UserID = ?", (user_id,))
    total_count = cursor.fetchone()[0]

    page_param = request.args.get('page')
    page, offset, total_pages = get_pagination(page_param, total_count, per_page)

    cursor.execute("""
        SELECT d.DocumentID, d.Title, d.Author, d.PublishDate
        FROM Favorites f
        INNER JOIN Documents d ON f.DocumentID = d.DocumentID
        WHERE f.UserID = ?
        ORDER BY f.CreateTime DESC
        OFFSET ? ROWS FETCH NEXT ? ROWS ONLY
    """, (user_id, offset, per_page))
    documents = cursor.fetchall()

    return render_template(
        'favorites.html',
        documents=documents,
        total_count=total_count,
        page=page,
        total_pages=total_pages
    )


@favorites_bp.route('/toggle_favorite', methods=['POST'])
def toggle_favorite():
    if 'user_id' not in session:
        return jsonify({'success': False, 'error': '未登录'})

    data = request.get_json()
    doc_id = data.get('document_id')
    user_id = session['user_id']

    if not doc_id:
        return jsonify({'success': False, 'error': '参数错误'})

    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("SELECT DocumentID FROM Documents WHERE DocumentID = ?", (doc_id,))
    if not cursor.fetchone():
        return jsonify({'success': False, 'error': '该文献不存在或已被删除'})

    cursor.execute("""
        SELECT FavoriteID FROM Favorites
        WHERE UserID = ? AND DocumentID = ?
    """, (user_id, doc_id))
    existing = cursor.fetchone()

    if existing:
        cursor.execute("DELETE FROM Favorites WHERE FavoriteID = ?", (existing[0],))
        is_favorited = False
    else:
        cursor.execute("INSERT INTO Favorites (UserID, DocumentID) VALUES (?, ?)", (user_id, doc_id))
        is_favorited = True

    conn.commit()

    return jsonify({'success': True, 'is_favorited': is_favorited})

File: static/js/favorites.js
// ============================================================
// static/js/favorites.js —— 我的收藏页交互逻辑
//
// 对应页面：favorites.html
// 功能：在收藏列表里点击"取消收藏"按钮，异步删除并刷新列表
// ============================================================


// ============================================================
// 取消收藏
//
// 参数：docId → 要取消收藏的文献 ID
//
// 流程：
// 1. 弹出确认对话框，用户点"取消"就中止操作
// 2. 向后端发送 POST 请求
// 3. 成功后刷新当前页面，更新收藏列表
// ============================================================
function cancelFavorite(docId) {
    // confirm 弹出一个浏览器原生的确认对话框
    // 用户点"确定"返回 true，点"取消"返回 false
    if (!confirm('确定要取消收藏这篇文献吗？')) {
        return;  // 用户点了"取消"，直接退出函数，什么都不做
    }

    fetch('/toggle_favorite', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ document_id: docId })
    })
    .then(res => res.json())
    .then(data => {
        if (data.success) {
            // 取消收藏成功后，刷新整个页面来更新列表
            // location.reload() 相当于按了浏览器的"刷新"按钮
            location.reload();
        }
    })
    .catch(err => {
        console.error('取消收藏失败：', err);
    });
}